distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/Rebalancer.kt

```diff
@@ -0,0 +1,261 @@
+package ru.fix.distributed.job.manager
+
+import org.apache.curator.framework.CuratorFramework
+import org.apache.curator.framework.imps.CuratorFrameworkState
+import org.apache.logging.log4j.kotlin.Logging
+import org.apache.zookeeper.KeeperException.NoNodeException
+import ru.fix.aggregating.profiler.Profiler
+import ru.fix.distributed.job.manager.model.AssignmentState
+import ru.fix.distributed.job.manager.model.JobId
+import ru.fix.distributed.job.manager.model.WorkItem
+import ru.fix.distributed.job.manager.model.WorkerId
+import ru.fix.distributed.job.manager.strategy.AssignmentStrategy
+import ru.fix.stdlib.concurrency.events.ReducingEventAccumulator
+import ru.fix.stdlib.concurrency.threads.NamedExecutors
+import ru.fix.zookeeper.transactional.ZkTransaction
+import ru.fix.zookeeper.utils.ZkTreePrinter
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+private const val ASSIGNMENT_COMMIT_RETRIES_COUNT = 3
+
+private val REBALANCE_EVENT = Any()
+
+internal class Rebalancer(
+        profiler: Profiler,
+        private val paths: ZkPathsManager,
+        private val curatorFramework: CuratorFramework,
+        private val leaderLatchExecutor: LeaderLatchExecutor,
+        private val assignmentStrategy: AssignmentStrategy,
+        private val nodeId: String
+) : AutoCloseable {
+    private val zkPrinter = ZkTreePrinter(curatorFramework)
+
+    private val rebalanceEventReceiverExecutor = NamedExecutors.newSingleThreadPool(
+            "rebalance-event-receiver", profiler)
+
+    private val rebalanceEventAccumulator = ReducingEventAccumulator.lastEventWinAccumulator<Any>()
+
+    fun handleRebalanceEvent() {
+        rebalanceEventAccumulator.publishEvent(REBALANCE_EVENT)
+    }
+
+    fun start() = rebalanceEventReceiverExecutor.execute {
+        rebalanceEventAccumulator.receiveReducedEventsUntilClosed {
+            leaderLatchExecutor.submitIfNeeded(Runnable(this::reassignAndBalanceTasks))?.get()
+        }
+    }
+
+    override fun close() {
+        rebalanceEventAccumulator.close()
+        rebalanceEventReceiverExecutor.shutdown()
+        if (!rebalanceEventReceiverExecutor.awaitTermination(3, TimeUnit.MINUTES)) {
+            logger.error("Failed to wait Rebalancer executor termination")
+            rebalanceEventReceiverExecutor.shutdownNow()
+        }
+    }
+
```

Comment:
here is below old rebalancing algorithm from original Manager.java with small refactoring